File: javaDS/Trees/BinaryTree.java
package javaDS.Trees;

public abstract class BinaryTree {
    
    protected Node root;
    protected int size;
    
    protected class Node<Data> {
        protected int key;
        protected Data data;
        protected Node parent = null;
        protected Node left = null;
        protected Node right = null;

        protected Node(int key, Data data) {
            this.key = key;
            this.data = data;
        }

        @Override
        public String toString() {
            String leftString;
            String rightString;

            if(this.left == null) {
                leftString = "Null";
            } else {
                leftString = this.left.toString();
            }

            if(this.right == null) {
                rightString = "Null";
            } else {
                rightString = this.right.toString();
            }

            return "{" + leftString + "} <-[" + this.key + "]-> {" + rightString + "}";
        }
    }

    protected BinaryTree() {}

    public int getSize() {
        return this.size;
    }

    public int getRoot() {
        return this.root.key;
    }

    private void inorder(Node current) {
        if(current != null) {
            inorder(current.left);
            System.out.println(current.key);
            inorder(current.right);
        }
    }
    
    public void inorder() {
        inorder(this.root);
    }

    private void preorder(Node current) {
        if(current != null) {
            System.out.println(current.key);
            preorder(current.left);
            preorder(current.right);
        }
    }
    
    public void preorder() {
        preorder(this.root);
    }

    private void postorder(Node current) {
        if(current != null) {
            postorder(current.left);
            postorder(current.right);
            System.out.println(current.key);
        }
    }
    
    public void postorder() {
        postorder(this.root);
    }

    private int maxHeight(int height1, int height2) {
        if(height1 >= height2) {
            return height1;
        } else {
            return height2;
        }
    }

    private int height(Node current) {
        if(current == null) {
            return 0;
        } else if(current.left == null && current.right == null) {
            return 0;
        }

        return maxHeight(height(current.left), height(current.right)) + 1;
    }

    public int height() {
        return height(this.root);
    }

    public int size() {
        return this.size;
    }
}

File: python/trees/binary_tree.py
class BinaryTree:
    
    size : int

    def __init__(self):
        self.root = None
        self.size = 0
    
    def __repr__(self):
        return f"{self.root}"

    def get_size(self):
        return self.size

    def get_root(self):
        return self.root.key
    
    def max_height(self, height1, height2):
        if height1 >= height2:
            return height1
        else:
            return height2    

    def __height(self, current):
        if not current:
            return 0
        elif not current.left and not current.right:
            return 0

        return self.max_height(self.__height(current.left), self.__height(current.right)) + 1

    def height(self):
        return self.__height(self.root)

    def __inorder(self, current):
        if current:
            self.__inorder(current.left)
            print(current.key)
            self.__inorder(current.right)
    
    def inorder(self):
        self.__inorder(self.root)

    def __preorder(self, current):
        if current:
            print(current.key)
            self.__preorder(current.left)
            self.__preorder(current.right)
    
    def preorder(self):
        self.__preorder(self.root)

    def __postorder(self, current):
        if current:
            self.__postorder(current.left)
            self.__postorder(current.right)
            print(current.key)
    
    def postorder(self):
        self.__postorder(self.root)